scripts/format_code.py

```diff
@@ -0,0 +1,257 @@
+
+# Parameters to append for all git diff executions. The ACMRT filter lists
+# the files that have been (A)dded (C)opied (M)odified (R)enamed or (T)ype
+# changed.
+GIT_DIFF_PARAMS = " --name-only --diff-filter=ACMRT"
+
+# Functions:    
+
+def does_file_need_formatting(filename):
+  """Executes clang-format on the file to determine if it includes any
+  formatting changes the user needs to fix.
+    
+  Returns:
+    True if the file requires format changes, False if formatting would produce
+    an identical file.
+  """
+  command_line = CLANG_FORMAT_BASE + CLANG_FORMAT_PARAM_LOG_REPLACEMENTS
+  command_line += filename
+  args = command_line.split(" ")
+  proc = subprocess.Popen(command_line.split(" "), stdout=subprocess.PIPE)
+  for line in io.TextIOWrapper(proc.stdout, encoding="utf-8"):
+    if line.strip().startswith("<replacement "):
+      return True
+  return False    
+
+def format_file(filename):
+  """Executes clang-format on the file to reformat it according to our
+  project's standards.  If -format_in_place is defined, the file will be
+  formatted in place, with the results overwriting the previous version
+  of the file. Otherwise the result of the file reformat will be logged
+  to standard out.
+
+  Args:
+   filename: path to the file to format.
+  """
+  command_line = CLANG_FORMAT_BASE + CLANG_FORMAT_PARM_IN_PLACE + filename
+  args = command_line.split(" ")
+  proc = subprocess.Popen(args)
+
+def git_diff_list_files():
+  """Compares the current branch to master to assemble a list of source
+  files which have been altered.
+  
+  Returns:
+    A list of file paths for each file in the git diff list with an extension
+    matching one of those in FILE_TYPE_EXTENSIONS.
+  """
+  filenames = []
+  command_line = (GIT_DIFF_BASE + FLAGS.git_range + GIT_DIFF_PARAMS).strip()
+  args = command_line.split(" ")
+  if FLAGS.verbose:
+    print()
+    print("Executing git diff to detect changed files, git_range:",
+      "\"" + FLAGS.git_range + "\"") 
+  proc = subprocess.Popen(args, stdout=subprocess.PIPE)
+  for line in io.TextIOWrapper(proc.stdout, encoding="utf-8"):
+    line = line.rstrip()
+    if(line.endswith(FILE_TYPE_EXTENSIONS)):
+      filenames.append(line.strip())
+      if FLAGS.verbose:
+        print("  - File detected: " + line)
+  if filenames and FLAGS.verbose:
+    print()
+  return filenames
+
+def list_files_from_directory(path, recurse):
+  """Iterates through the the directory and returns a list of files
+  which match those with extensions defined in FILE_TYPE_EXTENSIONS. 
+
+  Args:
+    path: the path to a directory to start searching from.
+    recurse: when True, will also recursively search for files in any
+      subdirectories found in path.
+  
+  Returns:
+    A list of matching filenames found in the directory.
+  """
+  filenames = []
+  for root, dirs, files in os.walk(path):
+    path = root.split(os.sep)
+    for filename in files:
+      if filename.endswith(FILE_TYPE_EXTENSIONS):
```

Comment:
endswith will not work with a tuple/list.

Reply:
```
_, ext = os.path.splitext(filename)
if ext in FILE_TYPE_EXTENSION:
```

Reply:
Actually it does work with tuples.   I tried it with a list and ran into the problem I think you are refering to, but making it a tuple smoothed it out.

Reply:
Whoa! TIL! Weirdly it doesn't work with lists but works with tuples.